### DISCUSSION TITLE
Short URL Service in TerriaJS

### DISCUSSION THREAD

**Discussion by arky** [TerriaJS/terriajs#5071] [Q&A] [answered]
TerriaJS seems to be using now discontinued [Google URL shortening service](https://developers.googleblog.com/2018/03/transitioning-google-url-shortener.html) in [modles/GoogleURLShortener.js](https://github.com/TerriaJS/terriajs/blob/master/lib/Models/GoogleUrlShortener.js)

Does anyone have suggestions on how to implement a short URL in TerriaJS. Does one have to use alternative service like [lstu.fr]
(https://lstu.fr/)

**Comment by hardreddata** [comment]
You can use s3 or GitHub gist when creating share links for maps.

See https://github.com/TerriaJS/terriajs-server/blob/master/serverconfig.json.example#L134-L169

This might be a good place to start, depending on your needs.

**Comment by tephenavies** [comment, marked as answer]
@arky We originally used Google URL shortener but our URLs got too long. At that point (a few years ago) we decided to instead save share data to JSON [from `terriajs-server`](https://github.com/TerriaJS/terriajs-server/blob/master/lib/controllers/share.js) and return a link that would allow reloading from that JSON. As Russell mentioned we support saving this JSON to Amazon S3 or GitHub Gist through `terriajs-server` out of the box. Additional providers should be fairly easy to add and we're happy to accept PRs to `terriajs-server` adding support for other JSON storage solutions.

**Reply by tephenavies** [reply]
To clarify: we advise against using URL shorteners to implement sharing as you'll hit a URL length limit somewhere when sharing, especially if you're sharing Stories.

**Reply by imakihi** [reply]
@steve9164 
We are trying to set up this URL shortening service. We set up productionserverconfig.json like following but it didn't work so far...

`{
  "port": 3001,
  "allowProxyFor": [
    "nicta.com.au",
    "gov.au",
    "csiro.au",
    "arcgis.com",
    "argo.jcommops.org",
    "www.abc.net.au",
    "geoserver.aurin.org.au",
    "mapsengine.google.com",
    "s3-ap-southeast-2.amazonaws.com",
    "adelaidecitycouncil.com",
    "www.dptiapps.com.au",
    "geoserver-123.aodn.org.au",
    "geoserver.imos.org.au",
    "nci.org.au",
    "static.nationalmap.nicta.com.au",
    "githubusercontent.com",
    "gov",
    "gov.uk",
    "gov.nz",
    "services.aremi.data61.io"
  ],
  "basicAuthentication": {
    "username": "test_viewer",
    "password": "pwd"
  },
  "shareUrlPrefixes": {
    "s": {
      "service": "s3",
      "region": "ap-northeast-1",
      "bucket": "test-terria-sharedata",
      "accessKeyId": "my access key id",
      "secretAccessKey": "my secret key",
      "keyLength": 16
    }
  }
}`

What we are missing? How can I make this work. 

Thank you!

**Reply by hardreddata** [reply]
My code link above missed the best part

```
    // Which service (of those defined above) should be used when new URLs are requested.
    newShareUrlPrefix: "s",
```

**Reply by imakihi** [reply]
@RussellGrew Thank you for the info. We could make it work!!
Hiroo

**Reply by geoffrowland** [reply]
Are there any suggestions for URL-shortening without terriajs-server, i.e running TerriaMap Static on GitHub Pages?

Thanks
